lib/solr.js

```diff
@@ -809,146 +692,218 @@ Client.prototype.createQuery = function () {
- * @param {Error} callback().err
- * @param {Object} callback().obj - JSON response sent by the Solr server deserialized
+ * Utility only used in tests.
+ *
+ * @param {string} fieldName
+ * @param {string} fieldType
+ *
+ * @returns {ClientRequest}
  *
- * @return {http.ClientRequest}
  * @api private
  */
+Client.prototype.createSchemaField = function(fieldName, fieldType) {
+  return this.doRequest(
+    this.getFullHandlerPath('schema'),
+    'POST',
+    pickJSON(this.options.bigint).stringify({
+      'add-field': {
+        name: fieldName,
+        type: fieldType,
+        multiValued: false,
+        stored: true,
+      },
+    }),
+    'application/json',
+    'application/json; charset=utf-8',
+    (err) => {
+      if (err) {
+        // TODO: We should handle this in a more robust way in the future, but
+        //   there is a difference between default setup in Solr 5 and Solr 8,
+        // so some fields already exist in Solr 8. Hence if that's the case, we
+        // just ignore that.
+        console.warn(err.message);
+      }
+    },
+  );
+};
 
-function postForm(params, callback) {
-  const headers = {
-    'content-type': 'application/x-www-form-urlencoded; charset=utf-8',
-    'content-length': Buffer.byteLength(params.params),
-    accept: 'application/json; charset=utf-8',
-  };
-  if (params.authorization) {
-    headers['authorization'] = params.authorization;
-  }
-  let options = {
-    host: params.host,
-    port: params.port,
-    method: 'POST',
-    headers: headers,
-    path: params.fullPath,
-    family: params.ipVersion,
-  };
-
-  if (params.request) {
-    options = Object.assign(options, params.request);
+/**
+ * Construct the full path to the given "handler".
+ *
+ * @param {string} handler
+ *   Relative URL path for the solr handler.
+ *
+ * @returns {string}
+ *   Full URL to the handler.
+ *
+ * @api private
+ */
+Client.prototype.getFullHandlerPath = function(handler) {
+  let pathArray;
+  if (handler === this.COLLECTIONS_HANDLER) {
+    pathArray = [this.options.path, handler];
+  } else {
+    pathArray = [
+      this.options.path,
+      this.options.core,
+      handler,
+    ];
   }
+  return pathArray.filter(e => e).join('/');
+};
 
-  if (params.agent !== undefined) {
-    options.agent = params.agent;
+/**
+ * Perform a query. Auto-detect whether GET or POST should be used.
+ *
+ * @param {string} handler
+ *   The solr "handler", i.e. request path.
+ * @param {Query | Collection | object | string} query
+ *   The query to perform.
+ * @param {Function} callback
+ *   The callback to run when the request completes.
+ *
+ * @returns {ClientRequest}
+ *
+ * @api private
+ */
+Client.prototype.doQuery = function(
+  handler,
+  query,
+  callback,
+) {
+  // Construct the string to use as query (GET) or body (POST).
+  let data;
+  if (query instanceof Query || query instanceof Collection) {
```

Comment:
This kind of highly conditional logic is confusing and very hard to read. I would vote in favour of keeping GET and POST requests separate and invoked explicitly.
Reducing amount of lines of code by itself is not necessarily a win, if that results in a spike of complexity.

Reply:
This conditional logic is not mine. I just moved it during the refactor. I agree that we should not do this, but I think it's a separate issue.